### PR TITLE
Integration test docs

### LINKED ISSUE #12178: Write proper integration test docs/handbook page
We currently have a very brief README that is already somewhat outdated. We need proper docs so other teams and new hires can build on our testing infrastructure.

The current docs also need to be updated to include integration tests in anything that before only applied to e2e tests.

### PR DESCRIPTION
**Pull request by felixfbecker**
Closes #12178.

Restructures, overhauls, and extends our testing docs for integration tests.
Some things for e2e tests were outdated, some apply to integration tests too. I tried to create a structure that makes sense (and will be shown in the automatic ToC on the right on docs.sourcegraph.com).

I also made some small DX improvements I noticed while writing/reading the existing the docs.

### CODE REVIEW

File: doc/dev/testing.md [hunk 1]
```diff
@@ -138,33 +235,29 @@ If the element you are trying to select appears multiple times on the page (e.g.
...
-page.click(".selector")
+page.click('.selector')
 ```
 
-### E2e caveats
-
-In the testing pyramid, e2e tests account for a small minority of all of the tests in an app. Only reach for e2e testing when it's too difficult to unit test something.
-
-In comparison to unit tests, e2e tests are slower and flakier but often more convenient.
+For more complex interactions, you can execute them atomically _within the browser_ using `page.evaluate()`.
```
Comment (lguychard) [inline doc/dev/testing.md:254]:
```suggestion
You can execute more complex interactions atomically _within the browser_ using `page.evaluate()`.
```

File: doc/dev/testing.md [hunk 2]
```diff
@@ -24,75 +28,82 @@ Usually while developing you will either have `yarn test --watch` running in a t
...
+- Timed out waiting for a selector to match because the CSS class in the web app changed: update the test code and implementation if the CSS selector is not a stable `test-*` identifier.
+- Timed out waiting for a selector to match because the page was still loading: use `waitForSelector(selector, { visible: true })`.
+- Page disconnected or browser session closed: another part of the test code might have called `page.close()` asynchronously, the browser crashed (check the video), or the build got canceled.
+- Node was detached from the DOM: components can change the DOM asynchronously, make sure to not rely on element handles.
+- Timing problems: Use `retry()` to "poll" for a condition that cannot be expressed through `waitForSelector()` (as opposed to relying on a fixed `setTimeout()`).
 
-If CI appears stuck on e2e tests, you can view the screen in [VNC Viewer](https://www.realvnc.com/en/connect/download/viewer/) (free) by forwarding port 5900 to the pod. Find the pod name on the top right of the step in Buildkite:
+Retrying the Buildkite step can help determine whether the test is flaky or broken. If it's flaky, disable it with `it.skip()` and file an issue on the author.
```
Comment (lguychard) [inline doc/dev/testing.md:98]:
```suggestion
Retrying the Buildkite step can help determine whether the test is flaky or broken. If it's flaky, [disable it with `it.skip()` and file an issue on the author](https://about.sourcegraph.com/handbook/engineering/testing#flaky-tests).
```

File: doc/dev/testing.md [hunk 3]
```diff
@@ -103,15 +114,99 @@ kubectl port-forward -n buildkite <buildkite agent pod> 5900:5900
...
+#### Writing integration tests
+
+Just like end-to-end tests, integration tests use the [test driver](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/shared/src/testing/driver.ts#L129:17) which is created for each test in a `before()` hook.
+In opposite to end-to-end tests, integration tests do not need to set up any backend state.
+Instead, integration tests create a _test context_ object before every test using `beforeEach()`, which manages the mocked responses.
 
-Open `web/src/e2e/e2e.test.ts` and add a new `test`:
+##### Mocking GraphQL responses
+
+Calling `testContext.overrideGraphQl()` in a test or `beforeEach()` hook with an object map allows you to override GraphQL queries and mutations that are made by the application during the test by their unique query name used by the implementation.
```
Comment (lguychard) [inline doc/dev/testing.md:141]:
```suggestion
Calling `testContext.overrideGraphQl()` in a test or `beforeEach()` hook with an object map allows you to override GraphQL queries and mutations made by the client code. The map is indexed by the unique query name specified in the implementation, for example [`ResolveRepo`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a3b40f3ae9376b42ce9a67b5a33f177ba98ac050/-/blob/browser/src/shared/repo/backend.tsx?subtree=true#L32-36).
```

File: doc/dev/testing.md [hunk 4]
```diff
@@ -103,15 +114,99 @@ kubectl port-forward -n buildkite <buildkite agent pod> 5900:5900
...
+The types of each query are specifically generated for each query to validate the shape of the mock results and provide autocompletion.
```
Comment (lguychard) [inline doc/dev/testing.md:142]:
```suggestion
Types for the variables and results of each query are generated, to validate the shape of the data at compile time and provide autocompletion.
```

File: doc/dev/testing.md [hunk 5]
```diff
@@ -103,15 +114,99 @@ kubectl port-forward -n buildkite <buildkite agent pod> 5900:5900
...
+If a query is made during a test that is not mocked, the request will be rejected and an error will be logged with details about which query was missing.
```
Comment (lguychard) [inline doc/dev/testing.md:143]:
```suggestion
If, during a test, a query is made that has no corresponding mock, the request will be rejected and an error will be logged with details about the query.
```

File: doc/dev/testing.md [hunk 6]
```diff
@@ -103,15 +114,99 @@ kubectl port-forward -n buildkite <buildkite agent pod> 5900:5900
...
+
+There are default mock responses for queries made in almost every test, which you can extend with object spread syntax if needed.
+
+##### Waiting for a mutation and checking passed variables
+
+To test the client made a GraphQL mutation to the backend, you can use `waitForRequest()`.
```
Comment (lguychard) [inline doc/dev/testing.md:149]:
```suggestion
To verify that the client sent a GraphQL mutation to the backend, you can use`waitForRequest()`.
```

File: doc/dev/testing.md [hunk 7]
```diff
@@ -103,15 +114,99 @@ kubectl port-forward -n buildkite <buildkite agent pod> 5900:5900
...
+The backend provides the webapp with a context object under `window.context`.
+You can override this object in integration tests using `testContext.overrideJsContext()`.
+There is a default mock JSContext that you can extend with object spread syntax if needed.
+
+### End-to-end tests
+
+End-to-end tests test the whole app: JavaScript, CSS styles, and backend.
+They can be found in [`web/src/end-to-end`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/web/src/end-to-end).
+
+A special end-to-end test suite is the regression test suite, which was created specifically for release testing and also contains some manual verification steps. As part of moving most of our current end-to-end tests to client & backend integration tests, the regression test suite will gradually be trimmed and phased out.
```
Comment (lguychard) [inline doc/dev/testing.md:167]:
```suggestion
The **regression test suite** is a special end-to-end test suite, which was created specifically for release testing and also contains some manual verification steps. As part of moving most of our current end-to-end tests to client & backend integration tests, the regression test suite will gradually be trimmed and phased out.
```

File: doc/dev/testing.md (hunk 4 above)
Comment (felixfbecker) [reply doc/dev/testing.md:142]: I wanted to express here that there is one type per query (now), as opposed to our previous one-size-fits-all types. Any suggestion to better communicate that?

Comment (felixfbecker) [reply doc/dev/testing.md:142]: Ah. How about replacing the first "each query" with "the overrides"?
```suggestion
The TypeScript types of the overrides are specifically generated for each query to validate the shape of the mock results and provide autocompletion.
```

Comment (lguychard) [reply doc/dev/testing.md:142]: Yup, that works.

**Pull request merged**